File: backend/test_qa_functionality.py
"""
Test script for basic Q&A functionality
This script tests the RAG chatbot API endpoints
"""

import asyncio
import aiohttp
import json

API_BASE_URL = "http://localhost:8000/api"  # Default, can be overridden

async def test_chatbot():
    print("Testing RAG Chatbot functionality...\n")

    async with aiohttp.ClientSession() as session:
        try:
            # Test 1: Create a new session
            print("1. Creating a new chat session...")
            async with session.post(
                f"{API_BASE_URL}/sessions",
                json={
                    "user_id": "test-user-123",
                    "metadata": {
                        "test": True,
                        "source": "test-script"
                    }
                }
            ) as response:
                if not response.ok:
                    raise Exception(f"Failed to create session: {response.status} {response.reason}")

                session_data = await response.json()
                session_id = session_data["session_id"]
                print(f"✓ Session created successfully: {session_id}")

            # Test 2: Send a sample question
            print("\n2. Sending a sample question...")
            question = "What is the main concept of Physical AI in humanoid robotics?"

            async with session.post(
                f"{API_BASE_URL}/sessions/{session_id}/messages",
                json={
                    "content": question,
                    "selected_text_context": None
                }
            ) as response:
                if not response.ok:
                    print(f"Warning: Failed to send message: {response.status} {response.reason}")
                    print("This may be expected if the vector store is not populated yet")
                    return

                message_data = await response.json()
                print("✓ Message sent successfully")
                print(f"Question: {question}")
                print(f"Response: {message_data['response'][:200]}...")

                if message_data.get("citations") and len(message_data["citations"]) > 0:
                    print(f"Citations found: {len(message_data['citations'])}")
                    for i, citation in enumerate(message_data["citations"], 1):
                        print(f"  {i}. {citation.get('chapter_name', 'Unknown')} ({citation.get('module_path', 'Unknown')})")
                else:
                    print("No citations returned (this may be expected if no relevant content was found)")

            # Test 3: Get message history
            print("\n3. Retrieving message history...")
            async with session.get(f"{API_BASE_URL}/sessions/{session_id}/messages") as response:
                if not response.ok:
                    raise Exception(f"Failed to get messages: {response.status} {response.reason}")

                history_data = await response.json()
                print(f"✓ Retrieved {len(history_data)} messages from history")

                # Display message history
                for i, message in enumerate(history_data, 1):
                    print(f"  {i}. [{message['role']}] {message['content'][:100]}...")

            # Test 4: Test with selected text context
            print("\n4. Testing with selected text context...")
            selected_text = "Physical AI is a new approach to robotics that focuses on the physical interaction between robots and their environment."
            follow_up_question = "Can you elaborate on this concept?"

            async with session.post(
                f"{API_BASE_URL}/sessions/{session_id}/messages",
                json={
                    "content": follow_up_question,
                    "selected_text_context": selected_text
                }
            ) as response:
                if not response.ok:
                    print(f"Warning: Failed to send context message: {response.status} {response.reason}")
                    print("This may be expected if the vector store is not populated yet")
                    return

                context_data = await response.json()
                print("✓ Context message sent successfully")
                print(f"Selected text: {selected_text}")
                print(f"Follow-up question: {follow_up_question}")
                print(f"Response: {context_data['response'][:200]}...")

            print("\n✓ All tests completed successfully!")
            print("\nSummary:")
            print("- Session creation: ✓ PASS")
            print("- Message sending: ✓ PASS")
            print("- Message history: ✓ PASS")
            print("- Context handling: ✓ PASS")

        except Exception as error:
            print(f"\n✗ Test failed with error: {str(error)}")
            return False

if __name__ == "__main__":
    asyncio.run(test_chatbot())

File: backend/test_qa_functionality.js
/**
 * Test script for basic Q&A functionality
 * This script tests the RAG chatbot API endpoints
 */

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:8000/api';

async function testChatbot() {
  console.log('Testing RAG Chatbot functionality...\n');

  try {
    // Test 1: Create a new session
    console.log('1. Creating a new chat session...');
    const sessionResponse = await fetch(`${API_BASE_URL}/sessions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        user_id: 'test-user-123',
        metadata: {
          test: true,
          source: 'test-script'
        }
      })
    });

    if (!sessionResponse.ok) {
      throw new Error(`Failed to create session: ${sessionResponse.status} ${sessionResponse.statusText}`);
    }

    const sessionData = await sessionResponse.json();
    const sessionId = sessionData.session_id;
    console.log('✓ Session created successfully:', sessionId);

    // Test 2: Send a sample question
    console.log('\n2. Sending a sample question...');
    const question = "What is the main concept of Physical AI in humanoid robotics?";

    const messageResponse = await fetch(`${API_BASE_URL}/sessions/${sessionId}/messages`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        content: question,
        selected_text_context: null
      })
    });

    if (!messageResponse.ok) {
      throw new Error(`Failed to send message: ${messageResponse.status} ${messageResponse.statusText}`);
    }

    const messageData = await messageResponse.json();
    console.log('✓ Message sent successfully');
    console.log('Question:', question);
    console.log('Response:', messageData.response.substring(0, 200) + '...');

    if (messageData.citations && messageData.citations.length > 0) {
      console.log('Citations found:', messageData.citations.length);
      messageData.citations.forEach((citation, index) => {
        console.log(`  ${index + 1}. ${citation.chapter_name} (${citation.module_path})`);
      });
    } else {
      console.log('No citations returned (this may be expected if no relevant content was found)');
    }

    // Test 3: Get message history
    console.log('\n3. Retrieving message history...');
    const historyResponse = await fetch(`${API_BASE_URL}/sessions/${sessionId}/messages`);

    if (!historyResponse.ok) {
      throw new Error(`Failed to get messages: ${historyResponse.status} ${historyResponse.statusText}`);
    }

    const historyData = await historyResponse.json();
    console.log('✓ Retrieved', historyData.length, 'messages from history');

    // Display message history
    historyData.forEach((message, index) => {
      console.log(`  ${index + 1}. [${message.role}] ${message.content.substring(0, 100)}...`);
    });

    // Test 4: Test with selected text context
    console.log('\n4. Testing with selected text context...');
    const selectedText = "Physical AI is a new approach to robotics that focuses on the physical interaction between robots and their environment.";
    const followUpQuestion = "Can you elaborate on this concept?";

    const contextResponse = await fetch(`${API_BASE_URL}/sessions/${sessionId}/messages`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        content: followUpQuestion,
        selected_text_context: selectedText
      })
    });

    if (!contextResponse.ok) {
      throw new Error(`Failed to send context message: ${contextResponse.status} ${contextResponse.statusText}`);
    }

    const contextData = await contextResponse.json();
    console.log('✓ Context message sent successfully');
    console.log('Selected text:', selectedText);
    console.log('Follow-up question:', followUpQuestion);
    console.log('Response:', contextData.response.substring(0, 200) + '...');

    console.log('\n✓ All tests completed successfully!');
    console.log('\nSummary:');
    console.log('- Session creation: ✓ PASS');
    console.log('- Message sending: ✓ PASS');
    console.log('- Message history: ✓ PASS');
    console.log('- Context handling: ✓ PASS');

  } catch (error) {
    console.error('\n✗ Test failed with error:', error.message);
    process.exit(1);
  }
}

// Run the test
testChatbot();